File: Others/PhoneKeypad.py


def word_numbers(input):
  digit_map = {
    '0' : '0',
    '1' : '1',
    '2': 'abc',
    '3': 'def',
    '4': 'ghi',
    '5': 'jkl',
    '6': 'mno',
    '7': 'pqrs',
    '8': 'tuv',
    '9': 'wxyz',
}
  input = str(input)
  ret = ['']
  for char in input:
   letters = digit_map.get(char,'')
   ret = [prefix+letter for prefix in ret for letter in letters]
  return ret

input = "2125420421"
result = []
result = word_numbers(input)
print result

File: Others/PhoneKeypad.rb
#!/usr/bin/ruby

def digit_map
 ['0', '1', 'abc', 'def', 'ghi', 'jkl', 'mno','pqrs','tuv','wxyz']
end

def word_numbers(phone_number)
  results = [""]
  letters = phone_number.split('')
  letters.each do |letter| # 2
    words = digit_map[letter.to_i] #abc
    results = results.inject([]) do |r, pre|
      r +=  words.split('').collect {|word| pre + word }
    end
  end
  results
end

p word_numbers("212")

# for prefix, each one add a new letter.
#   [] - > ["a", "b", "c"] -> ["a1", "b1", "c1"] -> ["a1a", "a1b", "a1c"...]
